janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/cache/CacheTransaction.java

```diff
@@ -105,10 +103,19 @@ public String toString() {
 
     private KCVMutation convert(KCVEntryMutation mutation) {
         assert !mutation.isEmpty();
-        if (!mutation.hasDeletions())
-            return new KCVMutation(mutation.getAdditions(), KeyColumnValueStore.NO_DELETIONS);
-        else
-            return new KCVMutation(mutation.getAdditions(), mutation.getDeletions().stream().map(KCVEntryMutation.ENTRY2COLUMN_FCT).collect(Collectors.toList()));
+        if (mutation.hasDeletions()) {
+            return new KCVMutation(
+                () -> new ArrayList<>(mutation.getAdditions()),
+                () -> {
+                    List<Entry> deletions = mutation.getDeletions();
+                    ArrayList<StaticBuffer> convertedDeletions = new ArrayList<>(deletions.size());
+                    for(Entry entry : deletions){
+                        convertedDeletions.add(KCVEntryMutation.ENTRY2COLUMN_FCT.apply(entry));
+                    }
+                    return convertedDeletions;
```

Comment:
Wouldn't it be more readable if we do:
```
return mutations.getDeletions()
    .stream()
    .map(KCVEntryMutation.ENTRY2COLUMN_FCT::apply)
    .collect(Collectors.toList());
```

Reply:
Oh this way we can even omit the code block and the return statement

Reply:
As noted and benchmarked by @mad , vanilla java is faster than stream api (https://github.com/JanusGraph/janusgraph/pull/2098#discussion_r415074443) and thus it is recommended to use vanilla java in performance critical sections (like mutations).
Personally, I didn't compare ArrayList transformation performance for vanilla java and stream API. Also, I'm not sure if StreamAPI is using initial ArrayList length to construct an array of needed size without resizing an array during additions. 
I think that StreamAPI is a cleaner version here, but I didn't check benchmarks. If you can confirm that Stream API uses `new ArrayList<>(initialArrayListSize)` instead of `new ArrayList<>()` for `Collectors.toList()` for stream from ArrayList and you checked performance and it is the same as vanilla java, I am more than happy to apply this suggestion. Otherwise, I think it would be better to use vanilla java code in this section.

Reply:
As far as I know, one can initialize a custom list to be used by `collect()`. But I have no benchmarks. Will do one if I have some time later today.

Reply:
@rngcntr indeed we could do `.collect(Collectors.toCollection(() -> new ArrayList<>(data.size())))`.
It would be good to check performance for this scenario. If it has the same performance, then we could use Steam API here.

Reply:
Using `ArrayList` with the stream API brings a slight improvement, but nothing beats vanilla:
```java
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@BenchmarkMode(Mode.AverageTime)
@OperationsPerInvocation(StreamVsVanilla.N)
@Warmup(iterations = 3, time = 2, timeUnit = TimeUnit.SECONDS)
@Measurement(iterations = 3, time = 2, timeUnit = TimeUnit.SECONDS)
public class StreamVsVanilla {
    public static final int N = 100_000;

    static List<String> sourceList = new ArrayList<>(N);
    static Function<String, String> identity = Function.identity();

    static {
        for (int i = 0; i < N; i++) {
            sourceList.add("hello");
        }
    }

    @Benchmark
    public List<String> vanilla() {
        List<String> adds = new ArrayList<>(N);
        for (final String s : sourceList) {
            adds.add(identity.apply(s));
        }
        return adds;
    }

    @Benchmark
    public List<String> stream() {
        List<String> adds = sourceList.stream()
            .map(identity)
            .collect(Collectors.toCollection(() -> new ArrayList<>(sourceList.size())));
        return adds;
    }

    @Benchmark
    public List<String> naiveStream() {
        List<String> adds = sourceList.stream()
            .map(identity)
            .collect(Collectors.toList());
        return adds;
    }
}
```
Results:
```
Benchmark                    Mode  Cnt   Score   Error  Units
StreamVsVanilla.naiveStream  avgt   15  11.339 ± 0.461  ns/op
StreamVsVanilla.stream       avgt   15  10.594 ± 0.548  ns/op
StreamVsVanilla.vanilla      avgt   15   6.362 ± 0.088  ns/op
```

Reply:
@rngcntr Thank you for these benchmarks. If so, we need to use vanilla code in performance critical sections. Also, I have created an issue here #2110 to move code to Vanilla Java in other performance critical sections.